ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c

```diff
@@ -0,0 +1,573 @@
+
+#define SAMP "ldms_geopm_sampler"
+#define GEOPM_METRIC_MAX 4096
+
+static ldmsd_msg_log_f msglog = NULL;
+
+/* Global state other than the logging function */
+static ldms_set_t g_set = NULL;
+static base_data_t g_base = NULL;
+static int g_metric_offset = 0;
+static int g_metric_cnt = 0;
+static int g_metric_pio_idx[GEOPM_METRIC_MAX] = {};
+static double g_metric[GEOPM_METRIC_MAX] = {};
+static enum ldms_geopm_sampler_metric_type_e g_metric_type[GEOPM_METRIC_MAX] = {};
+static char g_geopm_request_path[NAME_MAX] = "";
+
+
+enum ldms_value_type ldms_geopm_sampler_value_type(enum ldms_geopm_sampler_metric_type_e metric_type)
+{
+	enum ldms_value_type result;
+	switch(metric_type) {
+		case LDMS_GEOPM_SAMPLER_METRIC_TYPE_D64:
+			result = LDMS_V_D64;
+			break;
+		case LDMS_GEOPM_SAMPLER_METRIC_TYPE_S32:
+			result = LDMS_V_S32;
+			break;
+		case LDMS_GEOPM_SAMPLER_METRIC_TYPE_S64:
+			result = LDMS_V_S64;
+			break;
+		case LDMS_GEOPM_SAMPLER_METRIC_TYPE_U64:
+			result = LDMS_V_U64;
+			break;
+		default:
+			result = LDMS_V_NONE;
+			break;
+	}
+	return result;
+}
+
+int ldms_geopm_sampler_convert_sample(enum ldms_geopm_sampler_metric_type_e metric_type,
+				      double sample,
+				      union ldms_value *value)
+{
+	int rc = 0;
+	switch(metric_type) {
+		case LDMS_GEOPM_SAMPLER_METRIC_TYPE_D64:
+			value->v_d = sample;
+			break;
+		case LDMS_GEOPM_SAMPLER_METRIC_TYPE_S32:
+			value->v_s32 = (int32_t)sample;
+			break;
+		case LDMS_GEOPM_SAMPLER_METRIC_TYPE_S64:
+			value->v_s64 = (int64_t)sample;
+			break;
+		case LDMS_GEOPM_SAMPLER_METRIC_TYPE_U64:
+			value->v_u64 = geopm_signal_to_field(sample);
+			break;
+		default:
+			msglog(LDMSD_LERROR,
+			       SAMP": %s:%d: Error converting signal, unknown metric_type: %d\n",
+			       __FILE__, __LINE__, metric_type);
+			rc = EINVAL;
+			break;
+	}
+	return rc;
+}
+
+void ldms_geopm_sampler_set_log(void (*pf)(int, const char *fms, ...))
+{
+	msglog = pf;
+}
+
+int ldms_geopm_sampler_open_config(const char *config_path, FILE **result)
+{
+	int rc = 0;
+	*result = fopen(config_path, "r");
+	if (*result == NULL) {
+		msglog(LDMSD_LERROR,
+		       SAMP": %s:%d: Error in opening file '%s' containing list of samples, errno=%d\n",
+		       __FILE__, __LINE__, config_path, errno);
+		rc = ENOENT;
+	} else {
+		msglog(LDMSD_LDEBUG,
+		       SAMP": Success opening configuration file: '%s'\n",
+		       config_path);
+	}
+	return rc;
+}
+
+int ldms_geopm_sampler_parse_line(FILE *fid,
+				  char *signal_name,
+				  enum geopm_domain_e *domain_type,
+				  int *domain_idx,
+				  char *metric_name,
+				  enum ldms_geopm_sampler_metric_type_e *metric_type)
+{
+	char format[NAME_MAX];
+	int format_size = snprintf(format, sizeof(format),
+				   "%%%ds %%%ds %%d %%%ds ",
+				   NAME_MAX - 1, NAME_MAX - 1, NAME_MAX - 1);
+	if (format_size >= NAME_MAX) {
+		return ENAMETOOLONG;
+	}
+
+	char domain_type_str[NAME_MAX];
+	char *line = NULL;
+	size_t line_len = 0;
+	int num_scan = 0;
```

Comment:
set errno = 0; before depending on it as a check.

Reply:
Issue addressed. Let us know if it looks good now.